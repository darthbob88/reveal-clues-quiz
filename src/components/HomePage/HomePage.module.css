.question {
  border: 1px solid #bbb;
  margin-top: 1rem;
  padding: 0.5rem;
}
ul.clues {
  max-width: 60%;
  margin: 0 auto;
  list-style-position: inside;
}
button {
  display: block;
  margin: 0.5rem auto;
}
input {
  margin-left: 0.5rem;
}
