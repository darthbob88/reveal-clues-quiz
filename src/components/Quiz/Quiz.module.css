.question {
  border: 1px solid #bbb;
  margin-top: 1rem;
  padding: 0.5rem;
}
.disabled {
  background: #bcbcbc;
  color: #222;
}
.quiz {
  border: 1px solid lightgray;
  padding-bottom: 1em;
}
.startQuiz {
  opacity: 100%;
  display: inline-block;
  margin: 2rem 0.5rem 4rem;
}

div.nextPrevQs {
  display: flex;
  justify-content: center;
}
div.nextPrevQs button {
  margin: 0 1rem;
}
.questions li {
  display: inline-block;
  padding: 0.3rem;
  margin: 0.2rem;
  cursor: pointer;
}

.unanswered {
  background-color: #e6e6fa4f;
}
.correct {
  background-color: #7cfc007a;
}
.incorrect {
  background-color: #ff0000a6;
  color: white;
}

.current {
  border: 1px solid black;
}
