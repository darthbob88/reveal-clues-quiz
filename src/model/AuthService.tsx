import { firebaseAuth } from "../firebase";
import React, { useEffect, useState } from "react";
import { GoogleAuthProvider, User, createUserWithEmailAndPassword, sendPasswordResetEmail, signInWithEmailAndPassword, signInWithPopup } from "firebase/auth";

const doCreateUserWithEmailAndPassword = (
  email: string,
  password: string,
  userName: string
) =>
  createUserWithEmailAndPassword(firebaseAuth, email, password).then(result => {
    if (result.user == null) return false;
  }).catch(function (error) {
    console.log(error);
  });

const doSignInWithEmailAndPassword = (email: string, password: string) =>
  signInWithEmailAndPassword(firebaseAuth, email, password);

const doSignOut = () => firebaseAuth.signOut();

const doPasswordReset = (email: string) =>
  sendPasswordResetEmail(firebaseAuth, email);

// const doPasswordUpdate = (password: string) => {
//   if (firebaseAuth.currentUser != null) {
//     firebaseAuth.currentUser.updatePassword(password);
//   } else {
//     //TODO: If not logged in, throw an error?
//   }
// };

const doSignInWithGoogleOauth = () => {
  var provider = new GoogleAuthProvider();
  firebaseAuth.useDeviceLanguage();
  signInWithPopup(firebaseAuth, provider).then((result) => {
    // The signed-in user info.
    const user = result.user;
    console.log(user.displayName);
    // IdP data available using getAdditionalUserInfo(result)
    // ...
  }).catch((error) => {

  });

}


const currentUser: (User | null) = firebaseAuth.currentUser;
export const AuthService = {
  currentUser,
  doPasswordReset,
  doCreateUserWithEmailAndPassword,
  // doPasswordUpdate,
  doSignInWithEmailAndPassword,
  doSignInWithGoogleOauth,
  doSignOut
};

export const AuthContext = React.createContext(AuthService);

type AuthProviderTypes = { children: React.ReactNode };
export const AuthProvider: React.FunctionComponent<AuthProviderTypes> = ({
  children
}) => {
  const [currentUser, setCurrentUser] = useState<User | null>(null);
  useEffect(() => {
    firebaseAuth.onAuthStateChanged(user => setCurrentUser(user));
  });
  return (
    <AuthContext.Provider value={{ ...AuthService, currentUser }}>
      {children}
    </AuthContext.Provider>
  );
};
