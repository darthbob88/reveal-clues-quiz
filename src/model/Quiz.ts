import { action, computed, observable } from "mobx";
import { defaultQuestionState, Question, QuestionEnum, QuestionState } from "./Question";
/**
 * Overarching type for a quiz.
 * @field prompt: string A prompt for each question, to be used if there is no overruling one in the individual question
 * @field questions: Question[] A set of questions for each quiz.
 * @field title: An overarching title for the quiz. 
 */
export type Quiz = {
    questions: Question[];
    title: string;
    prompt: string;
}

export const defaultQuiz: Quiz = {
    title: "States by Attraction",
    prompt: "In what state will you find...",

    questions: [{
        clues: [
            "Ernest Hemingway’s Grave",
            "The Smurf Turf",
            "The Yellowstone Zone of Death",
            "The <State> Potato Museum"
        ],
        answer: "Idaho"
    },
    {
        clues: [
            "Weeki Wachee, the city of mermaids",
            "The Ernest Hemingway Home and Museum",
            "NASA Vehicle Assembly Building, a building so large it has its own weather",
            "The Southernmost Point in the Continental US"
        ],
        answer: "Florida"
    },
    {
        clues: [
            "World's Largest Collection of Smallest Versions of Largest Things",
            "Monument Rocks",
            "The Equality House, across the street from Westboro",
            "Geographic Center of the Contiguous United States"
        ],
        answer: "Kansas"
    },
    {
        clues: [
            "Mutter Museum of medical oddities",
            "Fallingwater, designed by Frank Lloyd Wright",
            "Gobbler’s Knob, home of Punxsutawney Phil",
            "The grave of Fred McFeely Rogers"
        ],
        answer: "Pennsylvania"
    },
    {
        clues: [
            "Slocum Mechanical Puzzle Collection",
            'Mentone Egg, A giant concrete egg marks this town as the "Egg Basket of the Midwest."',
            "Kokomantis",
            "Hoosier Hill"
        ],
        answer: "Indiana",
        revealOnAnswer: "https://www.atlasobscura.com/places/slocum-mechanical-puzzle-collection  https://www.atlasobscura.com/places/mentone-egg  https://www.atlasobscura.com/places/kokomantis-scrap-art-sculpture https://www.atlasobscura.com/places/hoosier-hill"
    },
    {
        clues: [
            "USPS Remote Encoding Facility",
            "Pando, the trembling giant",
            "The Spiral Jetty, a sculpture built onto a lake which is only visible during droughts",
            "Bonneville Salt Flats/Speedway"
        ],
        answer: "Utah"
    },
    {
        clues: [
            "The Very Large Array radiotelescope, outside of Socorro.",
            "Smokey Bear’s Grave",
            "Carlsbad Caverns",
            "The International UFO Museum and Research Center"
        ],
        answer: "New Mexico"
    },
    {
        clues: [
            "Vacuum Cleaner Museum",
            "Vista House at Crown Point",
            "Evergreen Aviation Museum, home of the Spruce Goose",
            "International Rose Test Garden, in the City of Roses"
        ],
        answer: "Oregon"
    },
    {
        clues: [
            "Rhea County Courthouse, site of the Scopes monkey trial",
            "Concrete Parthenon, built for the 1897 Centennial Exposition",
            "National Civil Rights Museum, established in the hotel where Dr Martin Luther King Jr was assassinated",
            "Jack Daniels Distillery"
        ],
        answer: "Tennessee"
    },
    {
        clues: [
            "Dr. Seuss House aka The Goose Creek Tower, a 185 foot tall log cabin",
            "The giant metal chicken statue, outside the town of Chicken",
            "St Michael’s Cathedral, first Russian Orthodox Church in North America",
            "Santa Claus House, in the town of North Pole"
        ],
        answer: "Alaska"
    },
    {
        clues: [
            "Ape Cave, longest continuous lava tube in the continental US",
            "Maryhill Museum and Stonehenge",
            "Brown and Haley factory store, manufacturers of Almond Roca",
            "Beacon Rock, one of the largest monoliths in the world"
        ],
        answer: "Washington"
    }]
}
export class QuizState {
    @observable currentQuiz: Quiz;
    // TODO: It'd be neat if I could make this part of the quiz itself
    @observable quizState: QuestionState[] = [];
    @observable timeRemaining: number = 0;
    @observable startedQuiz = false;


    constructor(chosenQuiz: Quiz) {
        this.currentQuiz = chosenQuiz;
        this.quizState = chosenQuiz.questions.map(question => ({ ...defaultQuestionState }));
        this.timeRemaining = 8 * 60 * 1000;
    }

    //TODO: Three methods, all called score*, is a little confusing.
    scoreQuestion(questionIndex: number, questionState: QuestionState) {
        if (this.quizState[questionIndex].state === QuestionEnum.UNANSWERED) {
            this.quizState[questionIndex] = questionState;
        }
    }

    @computed get isComplete() {
        const allQsAnswered = this.quizState.every(question => question.state !== QuestionEnum.UNANSWERED);
        const outOfTime = this.timeRemaining <= 0;
        return allQsAnswered || outOfTime;
    }

    @computed get scorePoints() {
        return this.quizState.reduce((acc, cur) => acc + cur.score, 0);
    }

    //TODO: fix this to return a string or smth, "right / total"
    @computed get scorePercent() {
        const numCorrect = this.quizState.filter(question => question.state === QuestionEnum.CORRECTLY_ANSWERED);
        return numCorrect.length;
    }

    startQuiz = () => {
        this.startedQuiz = true;
        this.measure();
    }

    @action measure() {
        if (!this.startedQuiz) return;

        this.timeRemaining -= 50;

        if (this.isComplete) {
            alert(`Time's up. Final score is ${this.scorePoints}`);
            return;
        }
        setTimeout(() => this.measure(), 50);
    }

    @computed get display() {
        const tenMilliSeconds = parseInt((this.timeRemaining / 10).toString(), 10);

        const seconds = parseInt((tenMilliSeconds / 100).toString(), 10);
        const minutes = parseInt((seconds / 60).toString(), 10);
        // TODO: Get some good method for formatting this for display.
        return `${minutes} : ${seconds % 60} :  ${tenMilliSeconds % 100}`;
    }
} 