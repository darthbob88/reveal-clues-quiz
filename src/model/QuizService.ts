import { Quiz, defaultQuiz, testQuizzes } from "./Quiz"
/**
 * Service for loading quizzes from whatever DB I decide to use.
 * Should probably make it parameterized, so I can select which quiz
 * to load, but that would mean handling saving quizzes, so.
 */

export const loadAllQuizzes = async (): Promise<Quiz[]> => {
    return Promise.resolve(testQuizzes);
}

export const loadQuiz = (slug: string) => {
    const selectedQuiz = testQuizzes.find(item => item.slug === slug) || defaultQuiz;
    return Promise.resolve(selectedQuiz);
}

export const saveScore = (user: string, score: number) => {
    localStorage.setItem(user, score.toString());
}