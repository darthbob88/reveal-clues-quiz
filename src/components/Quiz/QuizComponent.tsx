import React, { useState } from "react";
import { QuizState } from "../../model/Quiz";
import styles from "./Quiz.module.css";
import { QuestionComponent } from "../Question/Question";
import { QuestionEnum, QuestionState } from "../../model/Question";
import { observer } from "mobx-react";

type QuizProps = {
  quiz: QuizState;
};
export const QuizComp: React.FunctionComponent<QuizProps> = ({ quiz }) => {
  const [currentQuestion, setCurrentQuestion] = useState(0);
  const questions = quiz.currentQuiz.questions;

  const prevQuestion = () => {
    // This is stupid, but I can't think of a much better method
    let prev = questions.length + currentQuestion - 1;
    while (
      quiz.quizState[prev % questions.length].state !== QuestionEnum.UNANSWERED
    ) {
      prev--;
    }
    setCurrentQuestion(prev % questions.length);
  };

  const nextQuestion = () => {
    let next = currentQuestion + 1;
    while (
      quiz.quizState[next % questions.length].state !== QuestionEnum.UNANSWERED
    ) {
      next++;
    }
    setCurrentQuestion(next % questions.length);
  };

  const scoreQuestion = (points: QuestionState) => {
    quiz.scoreQuestion(currentQuestion, points);
    setTimeout(() => {
      nextQuestion();
    }, 800);
  };

  return (
    <div className={styles.quiz}>
      <p>
        Time remaining: {quiz.display} <br />
        Current score: {quiz.scorePoints} pts {quiz.scorePercent}/
        {quiz.quizState.length} question(s) correct
        {quiz.startedQuiz ? "" : <button className={styles.startQuiz} onClick={quiz.startQuiz}>Start Quiz</button>}
      </p>
      <div className={`${!quiz.startedQuiz ? styles.disabled : ""}`}>
        <ul className={styles.questions}>
          {quiz.quizState.map((question, index) => (
            <li
              key={index}
              onClick={() => setCurrentQuestion(index)}
              className={`${index === currentQuestion ? styles.current : ""}
              ${
                //TODO: This is ugly, but I can't find a good way to make this a mapping.
                question.state === QuestionEnum.UNANSWERED
                  ? styles.unanswered
                  : question.state === QuestionEnum.CORRECTLY_ANSWERED
                    ? styles.correct
                    : styles.incorrect
                }`}
            >
              {index + 1}
            </li>
          ))}
        </ul>
        <button onClick={() => prevQuestion()}>&lt; Previous Question</button>
        <button onClick={() => nextQuestion()}>Next Question &gt;</button>
        <QuestionComponent
          question={questions[currentQuestion]}
          state={quiz.quizState[currentQuestion]}
          awardPoints={scoreQuestion}
        />
      </div>
    </div>
  );
};

export const QuizComponent = observer(QuizComp);
