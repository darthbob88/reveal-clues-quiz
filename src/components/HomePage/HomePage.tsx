import React, { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import { Quiz } from "../../model/Quiz";
import * as QuizService from "../../model/QuizService";
import { ROUTES } from "../../routes";
import styles from "./HomePage.module.css";

type HomePageProps = {
  loadAllQuizzes?: () => Promise<Quiz[]>
}

export const HomePage: React.FunctionComponent<HomePageProps> = ({ loadAllQuizzes = QuizService.loadAllQuizzes }) => {
  const [quizList, setQuizList] = useState<Quiz[]>([]);
  useEffect(() => {
    let isSubscribed = true;
    loadAllQuizzes().then(result => {
      if (isSubscribed) {
        setQuizList(result);
      }
    });
    return () => { isSubscribed = false; };
  });


  return (
    <div className={styles.question}>
      <h2>List of Quizzes</h2>
      <ul className={styles.clues}>
        {quizList.length !== 0 ?
          quizList.map((quiz) => (
            <li key={quiz.slug}>
              {/* Trying to thread the needle between statically defined `ROUTES.HOME = "/home"``
            and hard-coding `/quiz/{quiz.slug}` everywhere I need it.
            Might have done better with a route-generation function, even if that would get clunky. */}
              <Link to={ROUTES.QUIZ.replace(":id", quiz.slug)}>{quiz.title}</Link>
            </li>
          ))
          : <p>Loading...</p>
        }
      </ul>
    </div>
  );
};
