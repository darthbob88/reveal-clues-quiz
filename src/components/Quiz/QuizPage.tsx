import React, { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { Quiz, QuizState } from "../../model/Quiz";
import { loadQuiz } from "../../model/QuizService";
import { QuizComponent } from "./QuizComponent";

export const QuizPage: React.FunctionComponent = () => {
    const { id } = useParams<{ id: string }>();
    const [quiz, setQuiz] = useState<Quiz>();
    useEffect(() => {
        const fetchData = async (id: string) => {
            const result = await loadQuiz(id);
            setQuiz(result);
        };
        id != null && fetchData(id);
    }, [id]);
    return (
        quiz === undefined ? <h2>Still Loading</h2> : <QuizComponent quiz={new QuizState(quiz)} />
    );
}