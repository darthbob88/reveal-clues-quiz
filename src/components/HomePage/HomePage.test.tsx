import React from "react";
import { act, render, waitForElement, waitForElementToBeRemoved } from "@testing-library/react";
import { HomePage } from "./HomePage";
import { MemoryRouter } from "react-router-dom";
import { testQuizzes } from "../../model/Quiz";
import { loadAllQuizzes } from "../../model/QuizService";
describe("Home page", () => {

  test("renders the home page properly", async () => {
    await act(async () => {
      const loadAllQuizzes = jest.fn().mockResolvedValue(testQuizzes);
      const { getByText, container } = render(<HomePage />, { wrapper: MemoryRouter });
      const linkElement = getByText(/List of Quizzes/i);
      expect(linkElement).toBeInTheDocument();
      expect(container).toMatchSnapshot();

      await waitForElementToBeRemoved(() => getByText('Loading...'))
    })

  });


  test("loads quizzes properly", async () => {
    await act(async () => {
      const loadAllQuizzes = jest.fn().mockResolvedValue(testQuizzes);
      const { getByText, container } = render(<HomePage />, { wrapper: MemoryRouter });
      const linkElement = getByText(/List of Quizzes/i);
      expect(linkElement).toBeInTheDocument();
      expect(container).toMatchSnapshot();

      await waitForElement(() => getByText(testQuizzes[0].title));
      getByText(testQuizzes[1].title);
    })

  });
})