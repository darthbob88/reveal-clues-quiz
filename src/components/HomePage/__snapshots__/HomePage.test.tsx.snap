// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page loads quizzes properly 1`] = `
<div>
  <div
    class="question"
  >
    <h2>
      List of Quizzes
    </h2>
    <ul
      class="clues"
    >
      <p>
        Loading...
      </p>
    </ul>
  </div>
</div>
`;

exports[`Home page renders the home page properly 1`] = `
<div>
  <div
    class="question"
  >
    <h2>
      List of Quizzes
    </h2>
    <ul
      class="clues"
    >
      <p>
        Loading...
      </p>
    </ul>
  </div>
</div>
`;
